docs/content/guides/migrations/from-step-launchers-to-pipes.mdx

```diff
@@ -0,0 +1,333 @@
+
+It's also possible to run Java or Scala Spark jobs with Dagster Pipes, but currently there is no official Pipes implementation for these languages. Therefore, forwarding Dagster events from these jobs is not yet supported officially (although it can be done with some custom code).
+
+### **2. Update your Dagster code**
+
+The goal is to keep the same observability and orchestration features while moving compute to an external script. Suppose you have existing code using Step Launchers similar to this:
+
+```python file=/guides/migrations/from_step_launchers_to_pipes/old_code.py
+from typing import Any
+
+from dagster_aws.emr import emr_pyspark_step_launcher
+from my_lib import MyPysparkIOManager, calculate_metric, get_data_frame
+from pyspark.sql import DataFrame
+
+import dagster as dg
+
+# the upstream asset will serve as an example of writing a Spark DataFrame
+
+
+@dg.asset(io_manager_key="pyspark_io_manager")
+def upstream(pyspark_step_launcher: dg.ResourceParam[Any]) -> DataFrame:
+    return get_data_frame()
+
+
+# the downstream asset will serve as an example of reading a Spark DataFrame
+# and logging metadata to Dagster
+
+
+@dg.asset(io_manager_key="pyspark_io_manager")
+def downstream(
+    context: dg.AssetExecutionContext,
+    upstream: DataFrame,
+    pyspark_step_launcher: dg.ResourceParam[Any],
+) -> None:
+    my_metric = calculate_metric(upstream)
+    context.add_output_metadata({"my_metric": my_metric})
+    return
+
+
+definitions = dg.Definitions(
+    assets=[upstream, downstream],
+    resources={
+        "pyspark_step_launcher": emr_pyspark_step_launcher,
+        "pyspark_io_manager": MyPysparkIOManager(),
+    },
+)
+```
+
+The corresponding Pipes code will be split in 2 pieces: the Dagster part and the PySpark job part.
+
+Let's start with the PySpark job. The upstream asset will invoke the following script:
+
+```python file=/guides/migrations/from_step_launchers_to_pipes/upstream_asset_script.py
+import os
+
+import boto3
+from dagster_pipes import PipesS3MessageWriter, open_dagster_pipes
+from my_lib import get_data_frame
+
+
+def main():
+    with open_dagster_pipes(
+        message_writer=PipesS3MessageWriter(client=boto3.client("s3")),
+    ) as pipes:
+        df = get_data_frame()
+
+        path = "s3://" + os.path.join(*pipes.asset_identifier) + ".parquet"
+
+        # this was probably previously logged by the IOManager
+        pipes.add_output_metadata({"path": path})
+
+        df.write.parquet(path)
+
+
+if __name__ == "__main__":
+    main()
+```
+
+Now, we have to run this script from Dagster. First, let's factor the boilerplate EMR config into a reusable function:
+
+```python file=/guides/migrations/from_step_launchers_to_pipes/utils.py startafter=start_emr_config_marker endbefore=end_emr_config_marker
+def make_emr_params(script_path: str) -> dict:
```

Comment:
It's nice to show a real-world example of using EMR, but in the earlier example we're not using EMR with the Step Launcher.

Would this example be simplified if we used local execution mode for Spark?

Reply:
We don't really have Pipes for local execution. 

> in the earlier example we're not using EMR with the Step Launcher.

We are, the EMR Step Launcher is passed to Definitions